tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,727 @@
+//! of permits at a time will always be woken eventually, even if many other
+//! tasks are acquiring smaller numbers of permits. This means that in a
+//! use-case like tokio's read-write lock, writers will not be starved by
+//! readers.
+//!
+//! The linked list is guarded by a mutex, which must be acquired before
+//! enqueueing or dequeueing a task. However, some operations are always
+//! wait-free.
+use crate::loom::{
+    cell::CausalCell,
+    sync::{atomic::AtomicUsize, Mutex, MutexGuard},
+};
+use crate::util::linked_list::{self, LinkedList};
+
+use std::{
+    cmp, fmt,
+    future::Future,
+    marker::PhantomPinned,
+    pin::Pin,
+    ptr::NonNull,
+    sync::atomic::Ordering::*,
+    task::{
+        Context, Poll,
+        Poll::{Pending, Ready},
+        Waker,
+    },
+};
+
+/// An asynchronous counting semaphore which permits waiting on multiple permits at once.
+pub(crate) struct Semaphore {
+    waiters: Mutex<Waitlist>,
+    /// The current number of available permits in the semaphore.
+    permits: AtomicUsize,
+
+    /// Permits in the process of being released.
+    ///
+    /// When releasing permits, if the lock on the semaphore's wait list is held
+    /// by another task, the task releasing permits adds them to this counter.
+    /// The task holding the lock will continue releasing permits until the
+    /// counter is drained, allowing the `release` operation to be wait free.
+    ///
+    /// The first bit of this counter indicates that the semaphore is closing.
+    /// Therefore, all values are shifted over one bit.
+    adding: AtomicUsize,
+}
+
+struct Waitlist {
+    queue: LinkedList<Waiter>,
+    closed: bool,
+}
+
+/// Error returned by `Permit::try_acquire`.
+#[derive(Debug)]
+pub(crate) enum TryAcquireError {
+    Closed,
+    NoPermits,
+}
+/// Error returned by `Permit::poll_acquire`.
+#[derive(Debug)]
+pub(crate) struct AcquireError(());
+
+#[derive(Debug)]
+pub(crate) struct Permit {
+    state: PermitState,
+}
+
+pub(crate) struct Acquire<'a> {
+    node: Waiter,
+    semaphore: &'a Semaphore,
+    permit: &'a mut Permit,
+    num_permits: u16,
+}
+
+/// Permit state
+#[derive(Debug, Copy, Clone)]
+enum PermitState {
+    /// Currently waiting for permits to be made available and assigned to the
+    /// waiter.
+    Waiting(u16),
+
+    /// The number of acquired permits
+    Acquired(u16),
+}
+
+/// An entry in the wait queue.
+struct Waiter {
+    /// The current state of the waiter. 
+    /// 
+    /// This is either the number of remaining permits required by
+    /// the waiter, or a flag indicating that the waiter is not yet queued.
+    state: AtomicUsize,
+
+    /// The waker to notify the task awaiting permits.
+    /// 
+    /// # Safety
+    /// 
+    /// This may only be accessed while the wait queue is locked.
+    /// XXX: it would be nice if we could enforce this...
+    waker: CausalCell<Option<Waker>>,
```

Comment:
Regarding enforcement of exclusive access, here's an idea: Introduce a zero-sized token that is obtained by calling a method on `Waitlist`, and require it for accessing the waker. Additionally put a lifetime on the token to ensure it doesn't outlive the mutex lock.
```rust
struct WaitlistToken<'a> {
    _p: PhantomData<&'a ()>,
}
impl Waitlist {
    #[inline]
    fn token<'a>(&'a self) -> WaitlistToken<'a> { ... }
}
struct TokenCell<T> {
    cell: CausalCell<T>,
}
impl<T> TokenCell<T> {
    fn with_mut<U, F: FnOnce(&T) -> U>(&self, _token: WaitlistToken, fn: F) -> U { ... }
}
```
As long as people don't generate the tokens with the struct initialization syntax (or ask another `Semaphore` for one), this guarantees the access happens while the mutex is locked. The token being zero-sized makes this zero-cost.

Reply:
I suppose you could also just pass a reference to the `Waitlist` to `with_mut` on the cell, instead of introducing a token.